gensim/models/poincare.py

```diff
@@ -0,0 +1,851 @@
+
+        self._gradients_computed = True
+
+    def compute_distance_gradients(self):
+        """Compute and store partial derivatives of poincare distance d(u, v) w.r.t all u and all v."""
+        if self._distance_gradients_computed:
+            return
+        self.compute_distances()
+
+        euclidean_dists_squared = self.euclidean_dists ** 2  # (1 + neg_size, batch_size)
+        c_ = (4 / (self.alpha * self.beta * np.sqrt(self.gamma ** 2 - 1)))[:, np.newaxis, :]  # (1 + neg_size, 1, batch_size)
+        u_coeffs = ((euclidean_dists_squared + self.alpha) / self.alpha)[:, np.newaxis, :]  # (1 + neg_size, 1, batch_size)
+        distance_gradients_u = u_coeffs * self.vectors_u - self.vectors_v  # (1 + neg_size, dim, batch_size)
+        distance_gradients_u *= c_  # (1 + neg_size, dim, batch_size)
+
+        nan_gradients = self.gamma == 1  # (1 + neg_size, batch_size)
+        if nan_gradients.any():
+            distance_gradients_u.swapaxes(1, 2)[nan_gradients] = 0
+        self.distance_gradients_u = distance_gradients_u
+
+        v_coeffs = ((euclidean_dists_squared + self.beta) / self.beta)[:, np.newaxis, :]  # (1 + neg_size, 1, batch_size)
+        distance_gradients_v = v_coeffs * self.vectors_v - self.vectors_u  # (1 + neg_size, dim, batch_size)
+        distance_gradients_v *= c_  # (1 + neg_size, dim, batch_size)
+
+        if nan_gradients.any():
+            distance_gradients_v.swapaxes(1, 2)[nan_gradients] = 0
+        self.distance_gradients_v = distance_gradients_v
+
+        self._distance_gradients_computed = True
+
+    def compute_loss(self):
+        """Compute and store loss value for the given batch of examples."""
+        if self._loss_computed:
+            return
+        self.compute_distances()
+
+        self.loss = -np.log(self.exp_negative_distances[0] / self.Z).sum()  # scalar
+        self._loss_computed = True
+
+
+class PoincareKeyedVectors(KeyedVectors):
+    """Class to contain vectors and vocab for the PoincareModel training class.
+
+    Used to perform operations on the vectors such as vector lookup, distance etc.
+
+    """
+    @staticmethod
+    def poincare_dist(vector_1, vector_2):
+        """Return poincare distance between two vectors."""
+        norm_1 = np.linalg.norm(vector_1)
+        norm_2 = np.linalg.norm(vector_2)
+        euclidean_dist = np.linalg.norm(vector_1 - vector_2)
+        if euclidean_dist == 0.0:
+            return 0.0
+        else:
+            return np.arccosh(
+                1 + 2 * (
+                    (euclidean_dist ** 2) / ((1 - norm_1 ** 2) * (1 - norm_2 ** 2))
+                )
+            )
+    # TODO: Add other KeyedVector supported methods - most_similar, etc.
+
+
+class PoincareRelations(object):
+    """Class to stream hypernym relations for `PoincareModel` from a tsv-like file."""
+
+    def __init__(self, file_path, encoding='utf8', delimiter='\t'):
+        """Initialize instance from file containing one hypernym pair per line.
```

Comment:
hypernym pair -> relation (here and elsewhere)

Reply:
Done